device.go

```diff
@@ -229,37 +231,174 @@ func virtioBlkDeviceHandler(_ context.Context, device pb.Device, spec *pb.Spec,
 }
 
-// updateSpecDeviceList takes a device description provided by the caller,
-// trying to find it on the guest. Once this device has been identified, the
-// "real" information that can be read from inside the VM is used to update
-// the same device in the list of devices provided through the OCI spec.
-// This is needed to update information about minor/major numbers that cannot
-// be predicted from the caller.
-func updateSpecDeviceList(device pb.Device, spec *pb.Spec) error {
-	// If no ContainerPath is provided, we won't be able to match and
-	// update the device in the OCI spec device list. This is an error.
-	if device.ContainerPath == "" {
+func vfioRebindDevice(bdf string) error {
+	fieldLogger := agentLog.WithField("guest-bdf", bdf)
+	fieldLogger.Debug("VFIO: Attempting to rebind")
+
+	sysPath := filepath.Join(sysBusPrefix, bdf)
+
+	// First we have to wait for the device to appear in the
+	// guest, since the hotplug can take a while.  We can't use
+	// the waiting logic in getDeviceName here, because the udev
+	// devices will only appear as we expect *after* we've
+	// rebound.
+	start := time.Now()
+	for {
+		_, err := os.Stat(sysPath)
+		if err == nil {
+			break
+		}
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("Error accessing %s: %s", sysPath, err)
+		}
+		if elapsed := time.Since(start); elapsed > vfioTimeout {
+			return fmt.Errorf("Timeout waiting for %s (%v)", sysPath, elapsed)
+		}
+		time.Sleep(vfioHotplugPollInterval)
+	}
+	fieldLogger.Debugf("VFIO: Found PCI device after %v", time.Since(start))
+
+	overridePath := filepath.Join(sysPath, "driver_override")
+	err := ioutil.WriteFile(overridePath, []byte("vfio-pci"), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", overridePath, err)
+	}
+
+	driverPath := filepath.Join(sysPath, "driver")
+	driver, err := os.Readlink(driverPath)
+	if os.IsNotExist(err) {
+		// This just means the device isn't bound to any driver
+		driver = ""
+	} else if err != nil {
+		return fmt.Errorf("Failed to readlink %s: %s", driverPath, err)
+	}
+
+	fieldLogger.Debugf("VFIO: Device previously bound to '%s'", driver)
+
+	if driver != "vfio-pci" && driver != "" {
+		fieldLogger.Debugf("VFIO: Unbinding from %s", driver)
+		unbindPath := filepath.Join(sysPath, bdf, "driver", "unbind")
+		err = ioutil.WriteFile(unbindPath, []byte(bdf), 0200)
+		if err != nil {
+			return fmt.Errorf("Failed to write %s: %s:", unbindPath, err)
+		}
+	}
+
+	err = ioutil.WriteFile(pciProbeFile, []byte(bdf), 0200)
+	if err != nil {
+		return fmt.Errorf("Failed to write %s: %s", pciProbeFile, err)
+	}
+
+	fieldLogger.Info("VFIO: Rebind complete")
+	return nil
+}
+
+func vfioDeviceHandler(ctx context.Context, device pb.Device, spec *pb.Spec, s *sandbox, devIdx devIndex) error {
+	fieldLogger := agentLog.WithField("host-dev", device.ContainerPath)
+
+	var group string
+
+	for _, opt := range device.Options {
+		l := strings.Split(opt, "=")
```

Comment:
You'll probably get complaints for the name `l`… Some other code in this file uses "tokens".

Reply:
Oh, for goodness sake.  It's a local, short is good.

Reply:
But, fine, whatever, I've renamed it to `tokens`.

Reply:
Thanks 😄